src/index.js

```diff
@@ -0,0 +1,161 @@
+import React, { PureComponent } from 'react';
+import { render } from 'react-dom';
+import PropTypes from 'prop-types';
+import { IntlProvider, addLocaleData, FormattedMessage } from 'react-intl';
+import NotInitialisedError from './NotInitialisedError';
+import supportedLanguages from './supportedLanguages.json';
+
+let translate = NotInitialisedError;
+let formatDate = NotInitialisedError;
+let Translation = NotInitialisedError;
+
+const withI18n = Component => {
+  class WithI18nComponent extends PureComponent {
+    render() {
+      return <Component {...this.props} translate={translate} formatDate={formatDate} />;
+    }
+  }
+
+  return WithI18nComponent;
+};
+
+class Provider extends PureComponent {
+  state = {
+    loaded: false,
+  };
+
+  async componentDidMount() {
+    const { domain } = this.props;
+    const language = this.getUserLanguage();
+    const localeData = this.getAllLocaleData();
+
+    addLocaleData(localeData);
+    const translations = await this.fetchTranslations(language);
+
+    const component = React.createElement(IntlProvider, {
+      messages: translations,
+      children: React.createElement('div'),
+      locale: language,
+      ref: element => {
+        const { intl } = element.getChildContext();
+
+        const provideIntlContext = Component => {
+          class IntlComponent extends PureComponent {
+            static childContextTypes = {
+              intl: PropTypes.object,
+            };
+
+            static propTypes = {
+              id: PropTypes.string.isRequired,
+            };
+
+            getChildContext() {
+              return { intl };
+            }
+
+            render() {
+              const { id, ...others } = this.props;
+              return <Component {...others} id={`${domain}.${id}`} />;
+            }
+          }
+
+          return IntlComponent;
+        };
+
+        translate = (id, values) => intl.formatMessage({ id: domain ? `${domain}.${id}` : id }, values);
+        formatDate = intl.formatDate;
+        Translation = provideIntlContext(FormattedMessage);
+
+        this.setState({ loaded: true });
+      },
+    });
+
+    render(component, document.createElement('div'));
+  }
+
+  getUserLanguage() {
+    if (this.props.isDevelopmentMode) {
+      return this.props.developmentLanguage;
+    }
+
+    const language = document.body.parentNode.getAttribute('lang');
+
+    if (!supportedLanguages.includes(language)) {
+      return this.props.fallbackLanguage;
+    }
+
+    return language;
+  }
```

Comment:
I'd prefer it if we would be able to simply pass the language as a prop (+ a `fallbackLanguage`) to determine which language should be used. I think `developmentMode` and `developmentLanguage` are things we shouldn't care about in this package but are rather things that the consumer of this component should decide when passing the `language` and `fallbackLanguage` props.

Then we also don't solely rely on the `html` DOM element having a `lang` attribute anymore making this more flexible to use. We could still use it though as a default if the `language` prop isn't set. (Also what do you think of using `document.documentElement` instead of `document.body.parentNode`?)

This would translate to something like this in code:
```
getUserLanguage() {
    const language = this.props.language || document.documentElement.getAttribute('lang');
    if (!supportedLanguages.includes(language)) {
        return this.props.fallbackLanguage;
    }

    return language;
}
```

I wonder wether it makes sense to have `fallbackLanguage` as a prop as it can still be overwritten to something that isn't supported. Perhaps we should hardcode it in this component to `en` or have a third hardcoded 'fallback language' in case the `fallbackLanguage` prop also isn't in the supported languages list.

Reply:
I was indeed thinking of dropping the development stuff. I'm going to drop it and do what you're suggesting here.

Reply:
Talked about this in real life to @lowiebenoot - makes sense @timdegroote 👍